### PR TITLE
Add encrypt/decrypt operations to key management interface

### PR DESCRIPTION
**Pull request by mikehelmick**
* add implementation in all supported key management services
  * AEAD is missing from azure keyvault, have reached out for help
* In support of #734

### CODE REVIEW

File: pkg/api/v1alpha1/exposure_types.go [hunk 1]
```diff
@@ -68,13 +72,24 @@ type Publish struct {
 	VerificationPayload  string        `json:"verificationPayload"`
 	HMACKey              string        `json:"hmackey"`
 	SymptomOnsetInterval int32         `json:"symptomOnsetInterval"`
+	RevisionToken        string        `json:"revisionToken"`
```
Comment (mikehelmick) [inline pkg/api/v1alpha1/exposure_types.go:75]: not implemented yet, showing API shape

File: pkg/keys/aws_kms.go [hunk 2]
```diff
@@ -49,3 +49,35 @@ func NewAWSKMS(ctx context.Context) (KeyManager, error) {
...
+func buildAAD(aad string) map[string]*string {
+	m := make(map[string]*string)
+	m["aad"] = &aad
+	return m
+}
+
+func (s *AWSKMS) Encrypt(ctx context.Context, keyID string, plaintext []byte, aad string) ([]byte, error) {
+	input := kms.EncryptInput{
+		KeyId:             &keyID,
+		EncryptionContext: buildAAD(aad),
```
Comment (sethvargo) [inline pkg/keys/aws_kms.go:62]: I think the AWS-ey Go way to do this is:

```go
map[string]*string{
  "aad": aws.String(aad),
}
```

File: pkg/keys/aws_kms.go [hunk 3]
```diff
@@ -49,3 +49,35 @@ func NewAWSKMS(ctx context.Context) (KeyManager, error) {
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to encrypt: %w", err)
+	}
+	return output.CiphertextBlob, nil
+}
+
+func (s *AWSKMS) Decrypt(ctx context.Context, keyID string, ciphertext []byte, aad string) ([]byte, error) {
+	input := kms.DecryptInput{
+		KeyId:             &keyID,
+		EncryptionContext: buildAAD(aad),
```
Comment (sethvargo) [inline pkg/keys/aws_kms.go:75]: Same thing

File: pkg/keys/hashicorp_vault.go [hunk 4]
```diff
@@ -54,20 +54,97 @@ func NewHashiCorpVault(ctx context.Context) (KeyManager, error) {
...
+
+func (v *HashiCorpVault) Encrypt(ctx context.Context, keyID string, plaintext []byte, aad string) ([]byte, error) {
+	kid, err := NewHCValueKeyID(keyID)
+	if err != nil {
+		return nil, err
+	}
+	pth := fmt.Sprintf("transit/encrypt/%s", kid.Name)
+	params := map[string]interface{}{
+		"plaintext": base64.StdEncoding.EncodeToString(plaintext),
+		"context":   base64.StdEncoding.EncodeToString([]byte(aad)),
```
Comment (sethvargo) [inline pkg/keys/hashicorp_vault.go:94]: We should document that this key needs to be created with `derived=true`

File: pkg/keys/hashicorp_vault.go [hunk 5]
```diff
@@ -54,20 +54,97 @@ func NewHashiCorpVault(ctx context.Context) (KeyManager, error) {
...
+		"type":      "aes256-gcm96",
+	}
+
+	result, err := v.client.Logical().Write(pth, params)
+	if err != nil {
+		return nil, fmt.Errorf("unable to encrypt: %w", err)
 	}
+
+	ciphertext, ok := result.Data["ciphertext"]
```
Comment (sethvargo) [inline pkg/keys/hashicorp_vault.go:103]: It's generally recommended that we store the entire ciphertext, including the `vault:vx:` prefix

File: pkg/keys/hashicorp_vault.go [hunk 6]
```diff
@@ -54,20 +54,97 @@ func NewHashiCorpVault(ctx context.Context) (KeyManager, error) {
...
+}
+
+func (v *HashiCorpVault) Decrypt(ctx context.Context, keyID string, ciphertext []byte, aad string) ([]byte, error) {
+	kid, err := NewHCValueKeyID(keyID)
+	if err != nil {
+		return nil, err
+	}
+	pth := fmt.Sprintf("transit/decrypt/%s", kid.Name)
+	params := map[string]interface{}{
+		"ciphertext": fmt.Sprintf("vault:v1:%s", base64.StdEncoding.EncodeToString(ciphertext)),
```
Comment (sethvargo) [inline pkg/keys/hashicorp_vault.go:128]: It won't always be v1 - that's why we should store the full blob that's returned.

File: pkg/keys/keys.go [hunk 7]
```diff
@@ -31,6 +31,9 @@ import (
 // KeyManager implementations must be able to return a crypto.Signer.
 type KeyManager interface {
 	NewSigner(ctx context.Context, keyID string) (crypto.Signer, error)
+
+	Encrypt(ctx context.Context, keyID string, plaintext []byte, aad string) ([]byte, error)
```
Comment (sethvargo) [inline pkg/keys/keys.go:35]: Can we doc these? Is it valid for aad to be ""?

Comment (sethvargo) [reply pkg/keys/keys.go:35]: Also, should aad be []byte?

File: pkg/keys/hashicorp_vault.go (hunk 4 above)
Comment (mikehelmick) [reply pkg/keys/hashicorp_vault.go:94]: added comments on the HashiCorpVault struct.

File: pkg/keys/hashicorp_vault.go (hunk 5 above)
Comment (mikehelmick) [reply pkg/keys/hashicorp_vault.go:103]: thanks for the tip.

File: pkg/keys/keys.go (hunk 7 above)
Comment (mikehelmick) [reply pkg/keys/keys.go]: went back and forth on it - agree []byte makes more sense.

practically, we'll probably take this in as a base64 env var

**Comment by mikehelmick** [comment]
rebased on main

**Pull request merged**